todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/domain/logic.kt

```diff
@@ -0,0 +1,38 @@
+@file:JvmName("TaskDetailLogic")
+
+package com.example.android.architecture.blueprints.todoapp.taskdetail.domain
+
+import com.example.android.architecture.blueprints.todoapp.data.Task
+import com.spotify.mobius.Effects.effects
+import com.spotify.mobius.Next
+import com.spotify.mobius.Next.*
+
+fun update(task: Task, event: TaskDetailEvent): Next<Task, TaskDetailEffect> =
+        when (event) {
+            TaskDetailEvent.DeleteTaskRequested -> dispatch(effects(TaskDetailEffect.DeleteTask(task)))
+            TaskDetailEvent.CompleteTaskRequested -> onCompleteTaskRequested(task)
+            TaskDetailEvent.ActivateTaskRequested -> onActivateTaskRequested(task)
+            TaskDetailEvent.EditTaskRequested -> dispatch(effects(TaskDetailEffect.OpenTaskEditor(task)))
+            TaskDetailEvent.TaskDeleted -> dispatch(effects(TaskDetailEffect.Exit))
+            TaskDetailEvent.TaskMarkedComplete -> dispatch(effects(TaskDetailEffect.NotifyTaskMarkedComplete))
+            TaskDetailEvent.TaskMarkedActive -> dispatch(effects(TaskDetailEffect.NotifyTaskMarkedActive))
+            TaskDetailEvent.TaskSaveFailed -> noChange()
+            TaskDetailEvent.TaskDeletionFailed -> noChange()
+        }
+
+private fun onActivateTaskRequested(task: Task): Next<Task, TaskDetailEffect> {
+    if (!task.details.completed) {
+        return noChange<Task, TaskDetailEffect>()
```

Comment:
Are the generics required here? Can't it be inferred from the return type?

Reply:
Yep, changed the call to just be `noChange()`